### DISCUSSION TITLE
taro 如何通过soucemap定位线上微信小程序问题

### DISCUSSION THREAD

**Discussion by zFitness** [NervJS/taro#15237] [Q & A] [answered]
目前公司的监控平台可以获取到小程序报错的堆栈信息以及微信官方小程序平台的sourcemap，但是无法通过行号定位到具体位置。
个人推测可能是：taro在build后会产生一次sourcemap，而上传build后的产物又会产生一个sourcemap，导致无法定位到源代码的位置。
所以想知道各位大佬如何解决的。

**Comment by jeffreyzhang23** [comment]
@zFitness 是的，微信上传会再生成一次SourceMap，通过微信CI工具虽然可以拿到第二次的SourceMap，但有很多重复路径，监控平台可能无法匹配到发生问题的文件。还是建议去微信后台定位源码位置，Taro只需开启SourceMap生成。

**Reply by zFitness** [reply]
但是错误的源码位置是第二次build后的，使用第二次的soucemap还原的代码位置不对。

**Comment by jeffreyzhang23** [comment]
> 但是错误的源码位置是第二次build后的，使用第二次的soucemap还原的代码位置不对。

不会，开发者工具会拿第一次SourceMap和源码一块处理，然后生成新的。

> 处理流程：源码 [ a.js a.js.map b.js b.js.map ] -> 开发者工具（JS转 ES5，压缩）-> 微信后台（合并 js 文件）[ appservice.app.js appservice.app.js.map]。

> 注意：如果源码在交给工具之前是经过了 webpack 等打包工具的处理，那源码这里需要有 map 文件。否则不需要存在 map 文件。

参见：https://developers.weixin.qq.com/community/minihome/article/doc/0008ee19284d9870524f9df4156013

**Reply by zFitness** [reply]
感谢，理解了。
但是有个问题，开发者工具第二次生成的sourcemap下载需要专业版才能使用，是否有第三方工具可以模拟生成sourcemap呢

**Comment by jeffreyzhang23** [comment, marked as answer]
> 感谢，理解了。 但是有个问题，开发者工具第二次生成的sourcemap下载需要专业版才能使用，是否有第三方工具可以模拟生成sourcemap呢

通过[微信CI工具](https://www.npmjs.com/package/miniprogram-ci)可以获取

**Reply by zFitness** [reply]
感谢，可以了。
现在我已经拿到合并后的soucemap，但是我发现一些问题，类似于：xxxx.appservice.js.map 这样的soucemap无法被 source-map-visualization 这个网站解析，报错如下：
<img width="774" alt="企业微信截图_e63da069-5d3b-41e5-955f-930df12520d4" src="https://github.com/NervJS/taro/assets/41500499/e9143ef7-2e5c-4795-8314-9df052a4e560">

而 pages/index/index.appservice.js 这种页面的soucemap可以正常解析。

**Reply by lanweifeng** [reply]
这个问题有解决方法吗？
